### PR TITLE
KIBA-1357: Token sales history

### PR DESCRIPTION
**Pull request by Ajadi-Abiola**
image
mobile:
image

### CODE REVIEW

File: app/src/theme.ts [hunk 1]
```diff
@@ -95,6 +95,7 @@ export const buildNotdTheme = (): ITheme => {
         // 'border-color': 'rgba(255, 255, 255, 0.5)',
         // 'border-style': 'solid',
         margin: '0',
+        'border-radius': '0',
```
Comment (krishan711) [inline app/src/theme.ts:98]: this seems wrong - you're changing all cards on the ui everywhere to have no radius?!

File: app/src/pages/TokenPage/TokenPage.tsx [hunk 2]
```diff
@@ -123,6 +138,66 @@ export const TokenPage = (): React.ReactElement => {
               <MetricView key={index} name={tokenAttribute.traitType} value={tokenAttribute.value} />
             ))}
           </Stack>
+          <Stack directionResponsive={{ base: Direction.Vertical, medium: Direction.Horizontal }} shouldWrapItems={true} isFullWidth={true} childAlignment={Alignment.Center} contentAlignment={Alignment.Center}>
+            <Text variant='header3'>Sales history</Text>
+            {tokenSales && tokenSales.map((tokenSale: TokenTransfer, index: number) : React.ReactElement => (
+              <Stack direction={Direction.Horizontal} key={index} shouldWrapItems={true} isFullWidth={true} childAlignment={Alignment.Center} contentAlignment={Alignment.Center}>
```
Comment (krishan711) [inline app/src/pages/TokenPage/TokenPage.tsx:144]: why do you have this stack?

File: app/src/pages/TokenPage/TokenPage.tsx [hunk 3]
```diff
@@ -123,6 +138,66 @@ export const TokenPage = (): React.ReactElement => {
...
+                      </Box>
+                    </Stack>
+                    <Stack direction={Direction.Horizontal} childAlignment={Alignment.Center} contentAlignment={Alignment.Center}>
+                      <Box variant='tokenTable' width='12rem' height='3rem'>
+                        <Account accountId={tokenSale?.fromAddress} />
+                      </Box>
+                      <Box variant='tokenTable' width='4rem' height='3rem'>
+                        <KibaIcon iconId='ion-arrow-forward' />
+                      </Box>
+                      <Box variant='tokenTable' width='12rem' height='3rem'>
```
Comment (krishan711) [inline app/src/pages/TokenPage/TokenPage.tsx:192]: this looks crazy - you've put the variant on lots of indiviudal boxes?! why not put all of this in one box doesnt that make sense?

File: app/src/pages/TokenPage/TokenPage.tsx [hunk 4]
```diff
@@ -20,6 +20,9 @@ export const TokenPage = (): React.ReactElement => {
 
   const [collectionToken, setCollectionToken] = React.useState<CollectionToken | undefined | null>(undefined);
   const [collection, setCollection] = React.useState<Collection | undefined | null>(undefined);
+  const [tokenSales, setTokenSales] = React.useState<TokenTransfer[] | undefined | null>(undefined);
+  const [offset, setOffset] = React.useState<number>(0);
```
Comment (krishan711) [inline app/src/pages/TokenPage/TokenPage.tsx:24]: you dont need to store this, just use tokenSales.length

File: app/src/client/client.ts [hunk 5]
```diff
@@ -48,6 +48,14 @@ export class NotdClient extends ServiceClient {
     return response.collectionStatistics;
   }
 
+  public getTokenRecentSales = async (registryAddress: string, tokenId: string, offset: number): Promise<Resources.TokenTransfer[]> => {
```
Comment (krishan711) [inline app/src/client/client.ts:51]: offset should be optional

File: app/src/client/endpoints.ts [hunk 6]
```diff
@@ -124,6 +124,35 @@ export class GetCollectionStatisticsResponse extends ResponseData {
     );
   }
 }
+export class GetTokenRecentSalesRequest extends RequestData {
+  readonly offset: number;
```
Comment (krishan711) [inline app/src/client/endpoints.ts:128]: offset should be optional

File: app/src/pages/TokenPage/TokenPage.tsx [hunk 7]
```diff
@@ -123,6 +145,76 @@ export const TokenPage = (): React.ReactElement => {
               <MetricView key={index} name={tokenAttribute.traitType} value={tokenAttribute.value} />
             ))}
           </Stack>
+          <Stack directionResponsive={{ base: Direction.Vertical, medium: Direction.Horizontal }} shouldWrapItems={true} isFullWidth={true} childAlignment={Alignment.Center} contentAlignment={Alignment.Center}>
+            <Text variant='header3'>Sales history</Text>
+            {tokenSales && tokenSales.map((tokenSale: TokenTransfer, index: number) : React.ReactElement => (
+              <Stack direction={Direction.Vertical} shouldWrapItems={true} key={index} isFullWidth={true} childAlignment={Alignment.Center} contentAlignment={Alignment.Center}>
```
Comment (krishan711) [inline app/src/pages/TokenPage/TokenPage.tsx:151]: please move this to its own component `TokenSaleRow`

File: app/src/client/endpoints.ts (hunk 6 above)
Comment (krishan711) [reply app/src/client/endpoints.ts:128]: @Ajadi-Abiola

File: app/src/client/endpoints.ts [hunk 8]
```diff
@@ -124,6 +124,35 @@ export class GetCollectionStatisticsResponse extends ResponseData {
...
+  constructor(offset?: number) {
+    super();
+    this.offset = offset || 0;
+  }
+  public toObject = (): Record<string, unknown> => {
```
Comment (krishan711) [inline app/src/client/endpoints.ts:133]: also, you need newlines between things in classes. when you write code please look at the code around it and match!

File: app/src/components/TokenSaleRow.tsx [hunk 9]
```diff
@@ -0,0 +1,80 @@
+import React from 'react';
+
+import { dateToString } from '@kibalabs/core';
+import { Alignment, Box, Direction, KibaIcon, PaddingSize, ResponsiveHidingView, ScreenSize, Spacing, Stack, Text, TextAlignment } from '@kibalabs/ui-react';
+
+import { TokenTransfer } from '../client/resources';
+import { Account } from './Account';
+
+export interface TokenSaleRowProps {
```
Comment (krishan711) [inline app/src/components/TokenSaleRow.tsx:9]: interfaces have an I at the start usually

File: app/src/components/TokenSaleRow.tsx [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+  tokenSale: TokenTransfer;
```
Comment (krishan711) [inline app/src/components/TokenSaleRow.tsx:10]: calle the variable tokenTransfer

File: app/src/pages/TokenPage/TokenPage.tsx [hunk 11]
```diff
@@ -56,6 +59,28 @@ export const TokenPage = (): React.ReactElement => {
...
+    }).catch((error: unknown): void => {
+      console.error(error);
+      setTokenSales(null);
+    });
+  }, [notdClient, registryAddress, tokenId]);
+  React.useEffect((): void => {
+    updateTokenSales();
+  }, [updateTokenSales]);
+
+  const onLoadClick = (): void => {
```
Comment (krishan711) [inline app/src/pages/TokenPage/TokenPage.tsx:75]: onLoadMoreClicked

File: app/src/pages/TokenPage/TokenPage.tsx [hunk 12]
```diff
@@ -56,6 +59,28 @@ export const TokenPage = (): React.ReactElement => {
...
+  const updateTokenSales = React.useCallback(async (): Promise<void> => {
+    setTokenSales(undefined);
+    notdClient.getTokenRecentSales(registryAddress, tokenId).then((tokenTransfers: TokenTransfer[]): void => {
+      setTokenSales(tokenTransfers);
+    }).catch((error: unknown): void => {
+      console.error(error);
+      setTokenSales(null);
+    });
+  }, [notdClient, registryAddress, tokenId]);
+  React.useEffect((): void => {
```
Comment (krishan711) [inline app/src/pages/TokenPage/TokenPage.tsx:71]: newline between functions

File: app/src/pages/TokenPage/TokenPage.tsx [hunk 13]
```diff
@@ -56,6 +59,28 @@ export const TokenPage = (): React.ReactElement => {
...
+    updateTokenSales();
+  }, [updateTokenSales]);
+
+  const onLoadClick = (): void => {
+    const tokenLength = tokenSales ? tokenSales.length : 0;
```
Comment (krishan711) [inline app/src/pages/TokenPage/TokenPage.tsx:76]: tokenSalesCount

File: app/src/theme.ts [hunk 14]
```diff
@@ -131,6 +128,13 @@ export const buildNotdTheme = (): ITheme => {
         'border-width': '0',
         'border-radius': '10px',
       },
+      tokenTable: {
```
Comment (krishan711) [inline app/src/theme.ts:131]: tokenSaleRow

File: app/src/client/client.ts (hunk 5 above)
Comment (Ajadi-Abiola) [reply app/src/client/client.ts]: done

**Review by krishan711 — APPROVED** [review]

File: app/src/client/endpoints.ts (hunk 6 above)
Comment (Ajadi-Abiola) [reply app/src/client/endpoints.ts]: done

**Pull request merged**